lib/client.js

```diff
@@ -33,151 +33,145 @@ const debug = require('debug')('knex:client');
 
 // The base client provides the general structure
 // for a dialect specific client object.
-function Client(config = {}) {
-  this.config = config;
-  this.logger = new Logger(config);
 
-  //Client is a required field, so throw error if it's not supplied.
-  //If 'this.dialect' is set, then this is a 'super()' call, in which case
-  //'client' does not have to be set as it's already assigned on the client prototype.
+class Client extends EventEmitter {
+  constructor(config = {}) {
+    super();
+    this.config = config;
+    this.logger = new Logger(config);
 
-  if (this.dialect && !this.config.client) {
-    this.logger.warn(
-      `Using 'this.dialect' to identify the client is deprecated and support for it will be removed in the future. Please use configuration option 'client' instead.`
-    );
-  }
-  const dbClient = this.config.client || this.dialect;
-  if (!dbClient) {
-    throw new Error(`knex: Required configuration option 'client' is missing.`);
-  }
-
-  if (config.version) {
-    this.version = config.version;
-  }
+    //Client is a required field, so throw error if it's not supplied.
+    //If 'this.dialect' is set, then this is a 'super()' call, in which case
+    //'client' does not have to be set as it's already assigned on the client prototype.
 
-  if (config.connection && config.connection instanceof Function) {
-    this.connectionConfigProvider = config.connection;
-    this.connectionConfigExpirationChecker = () => true; // causes the provider to be called on first use
-  } else {
-    this.connectionSettings = cloneDeep(config.connection || {});
-    this.connectionConfigExpirationChecker = null;
-  }
-  if (this.driverName && config.connection) {
-    this.initializeDriver();
-    if (!config.pool || (config.pool && config.pool.max !== 0)) {
-      this.initializePool(config);
+    if (this.dialect && !this.config.client) {
```

Comment:
I would suggest either removing this altogether or replacing warning with an error.

Reply:
I think this is not in scope of merge request, i will create later MR with throwing error